heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -44,36 +44,62 @@
   private final PriorityQueue<TimerTask> timers;
 
   // The tasks would be invoked before exit
-  private final ArrayList<Runnable> exitTasks;
+  private final List<Runnable> exitTasks;
 
   // For selector since there is bug in selector.select(timeout): we could not
   // use a timeout > 10 * Integer.MAX_VALUE
   // So here we set Integer.MAX_VALUE as the infinite future
   // We will also multiple 1000*1000 to convert mill-seconds to nano-seconds
   private static final Duration INFINITE_FUTURE = Duration.ofMillis(Integer.MAX_VALUE);
   private volatile boolean exitLoop;
+  private boolean terminateAllTasksOnWakeup;
+  private boolean terminateAllExitTasks;
 
   public WakeableLooper() {
     exitLoop = false;
-    tasksOnWakeup = new ArrayList<Runnable>();
+    tasksOnWakeup = new ArrayList<>();
     timers = new PriorityQueue<TimerTask>();
     exitTasks = new ArrayList<>();
+    terminateAllTasksOnWakeup = false;
+    terminateAllExitTasks = false;
   }
 
+  /**
+   * Clears all tasks.
+   * This is expected to be call synchronously.  This call is not thread safe!
```

Comment:
Can you remove this comment (and similar comments below), and add comments at the head to declare:
1. Most methods except the `onExit()` are not thread-safe.
2. People should handle the concurrent scenarios in their business logic rather than in this class.